File: src/main/java/org/reflections/vfs/SystemFile.java
package org.reflections.vfs;

import java.io.*;

/**
 * an implementation of {@link org.reflections.vfs.Vfs.File} for a directory {@link java.io.File}
 */
public class SystemFile implements Vfs.File {

    private final SystemDir    root;
    private final java.io.File file;

    public SystemFile(SystemDir root, java.io.File file) {
        this.root = root;
        this.file = file;
    }

    @Override
    public String getName() {
        return file.getName();
    }

    @Override
    public String getRelativePath() {
        String filepath = file.getPath().replace("\\", "/");
        if (filepath.startsWith(root.getPath())) {
            return filepath.substring(root.getPath().length() + 1);
        }

        return null; //should not get here
    }

    @Override
    public InputStream openInputStream() {
        try {
            return new FileInputStream(file);
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public String toString() {
        return file.toString();
    }
}

File: src/main/java/org/reflections/vfs/SystemFile.kt
package org.reflections.vfs

import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.InputStream

/**
 * an implementation of [org.reflections.vfs.Vfs.File] for a directory [java.io.File]
 */
class SystemFile(private val root: SystemDir, private val file: java.io.File) : Vfs.File {

    override val name: String
        get() {
            return file.name
        }
    override val relativePath: String?
        get() {
            val filepath = file.path.replace("\\", "/")
            return if (filepath.startsWith(root.path)) {
                filepath.substring(root.path.length + 1)
            } else null

            //should not get here
        }

    override fun openInputStream(): InputStream {
        try {
            return FileInputStream(file)
        } catch (e: FileNotFoundException) {
            throw RuntimeException(e)
        }

    }

    override fun toString(): String {
        return file.toString()
    }
}
